### PR TITLE
Multiple fixes for Windows x86 p/invoke test failures. 

### PR DESCRIPTION
**Pull request by lateralusX**
Fix for pinvoke2::test_0_marshal_empty_struct:

Fix for non GCC win x86 calling convention when passing empty structs over pinvoke. Port of https://github.com/mono/mono/pull/3085 to Windows x86.

Fix for pinvoke3::test_0_marshal_small_struct_delegate9:

Fix for incorrect implementation of small struct (single float/double) on Windows x86. Windows x86 ABI for returning structs of size 4 or 8 bytes (regardless of type) dictates that values are passed in EDX:EAX register pairs, https://msdn.microsoft.com/en-us/library/984x0h58.aspx. This is different compared to for example float or double return types (not in struct) that will be returned in ST(0),     https://msdn.microsoft.com/en-us/library/ha59cbfz.aspx.

Fix disables the alternative to do special handling of structs with 1 float or double and make sure same logic is always applied to all small structs regardless type as dictated by Windows x86 ABI.

Fix for pinvoke3::pinvoke3::test_55_marshal_delegate_ref_delegate:

mono_test_marshal_delegate_ref_delegate uses incorrect calling convention. mono_test_marshal_delegate_ref_delegate called by invoke3::test_55_marshal_delegate_ref_delegate uses incorrect calling convention. Mono assumes delegates to be stdcall but current definitions will be cdecl, this will cause an incorrect stack pointer on architectures where cdecl and stdcall are different (Windows x86). Test fails with "corrupt" stack pointer on Windows x86.

### CODE REVIEW

**Comment by lewurm** [comment]
build

File: mono/mini/mini-x86.h [hunk 1]
```diff
@@ -315,6 +315,7 @@ typedef struct {
 	/* Only if storage == ArgValuetypeInReg */
 	ArgStorage pair_storage [2];
 	gint8 pair_regs [2];
+	gboolean pass_empty_struct; // Set in scenarios when empty structs needs to be represented as argument.
```
Comment (kumpera) [inline mono/mini/mini-x86.h]: We should convert all gboolean fields to use a bitfield and avoid wasting space.

File: mono/mini/mini-x86.c [hunk 2]
```diff
@@ -292,7 +324,7 @@ add_valuetype (MonoMethodSignature *sig, ArgInfo *ainfo, MonoType *type,
  * For x86 ELF, see the "System V Application Binary Interface Intel386 
  * Architecture Processor Supplment, Fourth Edition" document for more
  * information.
- * For x86 win32, see ???.
+ * For x86 win32, see https://msdn.microsoft.com/en-us/library/984x0h58.aspx.
```
Comment (xmcclure) [inline mono/mini/mini-x86.c]: #nice

**Comment by lateralusX** [comment]
Adjusted PR based on feedback. OK to merge?

**Comment by migueldeicaza** [comment]
Looks good to me, @lateralusX go ahead and merge

**Pull request merged**